pkg/autoscaler/autoscaler.go

```diff
@@ -254,295 +256,406 @@ var _ = Describe("[Feature:Machines] Autoscaler should", func() {
+			for i := range machineSets {
+				scaledGroups[path.Join(machineSets[i].Namespace, machineSets[i].Name)] = false
+			}
+			scaleUpCounter := newScaleUpCounter(eventWatcher, 0, scaledGroups)
+			maxNodesTotalReachedCounter := newMaxNodesTotalReachedCounter(eventWatcher, 0)
+			// +1 to continuously generate the MaxNodesTotalReached
+			workload := newWorkLoad(int32(maxNodesTotal+1), workloadMemRequest, autoscalerWorkerNodeRoleLabel)
+			Expect(client.Create(ctx, workload)).Should(Succeed())
+			cleanupObjects[workload.Name] = workload
+			testDuration = time.Now().Add(time.Duration(framework.WaitLong))
+			Eventually(func() bool {
+				v := scaleUpCounter.get()
+				glog.Infof("[%s remaining] Expecting %v %q events; observed %v",
+					remaining(testDuration), clusterExpansionSize-1, clusterAutoscalerScaledUpGroup, v)
+				return v == uint32(clusterExpansionSize-1)
+			}, framework.WaitLong, pollingInterval).Should(BeTrue())
+
+			// The cluster-autoscaler can keep on generating
+			// ScaledUpGroup events but in this scenario we are
+			// expecting no more as we explicitly capped the
+			// cluster size with maxNodesTotal (i.e.,
+			// clusterExpansionSize -1). We run for a period of
+			// time asserting that the cluster does not exceed the
+			// capped size.
+			testDuration = time.Now().Add(time.Duration(framework.WaitShort))
+			Eventually(func() uint32 {
+				v := maxNodesTotalReachedCounter.get()
+				glog.Infof("[%s remaining] Waiting for %s to generate a %q event; observed %v",
+					remaining(testDuration), clusterAutoscalerComponent, clusterAutoscalerMaxNodesTotalReached, v)
+				return v
+			}, framework.WaitShort, pollingInterval).Should(BeNumerically(">=", 1))
+
+			testDuration = time.Now().Add(time.Duration(framework.WaitShort))
+			Consistently(func() bool {
+				v := scaleUpCounter.get()
+				glog.Infof("[%s remaining] At max cluster size and expecting no more %q events; currently have %v, max=%v",
+					remaining(testDuration), clusterAutoscalerScaledUpGroup, v, clusterExpansionSize-1)
+				return v == uint32(clusterExpansionSize-1)
+			}, framework.WaitShort, pollingInterval).Should(BeTrue())
+
+			By("Deleting workload")
+			scaleDownCounter := newScaleDownCounter(eventWatcher, uint32(clusterExpansionSize-1))
+			Expect(deleteObject(workload.Name, cleanupObjects[workload.Name])).Should(Succeed())
+			delete(cleanupObjects, workload.Name)
+			testDuration = time.Now().Add(time.Duration(framework.WaitLong))
+			Eventually(func() bool {
+				v := scaleDownCounter.get()
+				glog.Infof("[%s remaining] Expecting %v %q events; observed %v",
+					remaining(testDuration), clusterExpansionSize-1, clusterAutoscalerScaleDownEmpty, v)
+				return v == uint32(clusterExpansionSize-1)
+			}, framework.WaitLong, pollingInterval).Should(BeTrue())
+
+			Eventually(func() bool {
+				podList := corev1.PodList{}
+				err = client.List(ctx, &podList, runtimeclient.InNamespace(workload.Namespace))
+				Expect(err).NotTo(HaveOccurred())
```

Comment:
Calling `Expect` within an `Eventually` will bail if the whole test if there's an error, is that intentional or would we want to retry?